sql/src/main/java/org/apache/druid/sql/http/SqlResource.java

```diff
@@ -320,4 +181,187 @@ public Response cancelQuery(
       return Response.status(Status.FORBIDDEN).build();
     }
   }
+
+  /**
+   * The SqlResource only generates metrics and doesn't keep track of aggregate counts of successful/failed/interrupted
+   * queries, so this implementation is effectively just a noop.
+   */
+  private static class SqlResourceQueryMetricCounter implements QueryResource.QueryMetricCounter
+  {
+    @Override
+    public void incrementSuccess()
+    {
+
+    }
+
+    @Override
+    public void incrementFailed()
+    {
+
+    }
+
+    @Override
+    public void incrementInterrupted()
+    {
+
+    }
+
+    @Override
+    public void incrementTimedOut()
+    {
+
+    }
+  }
+
+  private class SqlResourceQueryResultPusher extends QueryResultPusher
+  {
+    private final String sqlQueryId;
+    private final HttpStatement stmt;
+    private final SqlQuery sqlQuery;
+
+    public SqlResourceQueryResultPusher(
+        AsyncContext asyncContext,
+        String sqlQueryId,
+        HttpStatement stmt,
+        SqlQuery sqlQuery
+    )
+    {
+      super(
+          (HttpServletResponse) asyncContext.getResponse(),
+          SqlResource.this.jsonMapper,
+          SqlResource.this.responseContextConfig,
+          SqlResource.this.selfNode,
+          SqlResource.QUERY_METRIC_COUNTER,
+          sqlQueryId,
+          MediaType.APPLICATION_JSON_TYPE
+      );
+      this.sqlQueryId = sqlQueryId;
+      this.stmt = stmt;
+      this.sqlQuery = sqlQuery;
+    }
+
+    @Override
+    public ResultsWriter start()
+    {
+      return new ResultsWriter()
+      {
+        private ResultSet thePlan;
+
+        @Override
+        @Nullable
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        public QueryResponse<Object> start(HttpServletResponse response)
+        {
+          response.setHeader(SQL_QUERY_ID_RESPONSE_HEADER, sqlQueryId);
+
+          final QueryResponse<Object[]> retVal;
+          try {
+            thePlan = stmt.plan();
+            retVal = thePlan.run();
```

Comment:
This is OK... But I worry that we create an opportunity for a naive user to try to use the statement both here and in the request. Nothing in the SQL planner is multi-threaded: each operation must be called by only a single thread at a time. (It is obviously OK for the thread ownership to change, as here.)

Also, in JDBC, there is an elaborate mechanism because the thread that starts the query (metrics?) must be the one that finishes them. I don't know if that restriction is still valid. If it is, this solution should be better than the original code.

I wonder if it is better to do the planing in the request thread, and just return an error response if things go south. Only if the plan is good do we move onto the execution being done here.

Reply:
Fwiw, this refactor doesn't actually change any thread ownership at all.  All of the work that was previously being done by the request thread is still doing that.  The RequestResultsPusher could theoretically be used across thread boundaries and the API now might make it look a bit more tempting to do that, but we should be careful if/when we actually do that.